discourse/discourse
# FEATURE: Improving bookmarks part 2 -- Lists and Topics

This PR makes quite a few changes, but it can be safely merged because all the changes are gated between the `enable_bookmarks_with_reminders` setting. If it is off none of this new stuff will happen.

The aim of this PR is to get the new bookmarks feature in a state where we can turn the setting on and it just works as a replacement for the old bookmarks, and then we can test in dev and meta. Then it's just a matter of adding the background reminder functionality.

### Testing

* Enable `SiteSetting.enable_bookmarks_with_reminders`.
* Test topic-level bookmarking and interaction with post bookmarks & clearing all bookmarks for topic.
* Go to user Activity tab and click Bookmarks, make sure it goes to the new list.
* Test deleting bookmarks from the new list and that clicking the link opens the correct post. Also check the name of the bookmark (if given) shows here.

### Main Changes

* Added a rake task to sync the old PostAction bookmarks to the new Bookmark table, which can be run as many times as we want for a site (it will not create duplicates).
* Changed the bookmarks topic filter at /bookmarks to load new bookmark records. This involved changing several topic serializers. This will show multiple records per topic because each bookmarked post can have its own reminder settings and name. Bookmarks can be deleted from here from the context menu. The icon changes based on whether the bookmark has a reminder attached. I had to make my own bookmark-topic-list-item.hbs template here to do the customization I needed; if there is a better way I am all ears.
* Disabled selecting the "reminder type" for bookmarks in the UI because the backend functionality is not done yet (of sending users notifications etc.)
* Hook up topic-level bookmarking to work with the new Bookmark table.

![image](https://user-images.githubusercontent.com/920448/73815517-f3da7880-4831-11ea-98ec-73f65a7039e9.png)

### Not Addressed

I will do the following in another PR:

* I have not yet changed the reports for bookmarks at /admin/reports/bookmarks.
* I have not made the topic-level bookmarks button trigger the new reminder modal.
